### DISCUSSION TITLE
[typescript] How to properly type `useRef` generic with AlertDialog, so it and the button won't complain?

### DISCUSSION THREAD

**Discussion by RIP21** [chakra-ui/chakra-ui#2936] [Q&A] [answered]
Hey folks. Thanks for Charkra, it's an amazing product! Although I finally hit the roadblock where I need some help :) 

How to properly type `useRef` so it won't complain? Or maybe it's an imperfection of typings that shall be fixed?
Here is how I tried to tackle it, but with no success. Union or `&` in generic of just `FocusableElement` from `@chakra/utils` doesn't help.
```typescript
export const DeleteModal = () => {

  const cancelRef = useRef<HTMLButtonElement | FocusableElement>()

  return (
    <AlertDialog isOpen leastDestructiveRef={cancelRef} onClose={onClose}>
      <AlertDialogOverlay>
        <AlertDialogContent>
          <AlertDialogHeader fontSize="lg" fontWeight="bold">
            Delete smth
          </AlertDialogHeader>

          <AlertDialogBody>
            Are you sure? You can&apos;t undo this action afterwards.
          </AlertDialogBody>

          <AlertDialogFooter>
            <Button ref={cancelRef} onClick={onClose}>
              Cancel
            </Button>
            <Button colorScheme="red" onClick={onClose} ml={3}>
              Delete
            </Button>
          </AlertDialogFooter>
        </AlertDialogContent>
      </AlertDialogOverlay>
    </AlertDialog>
  )
}
```
I get two errors on `AlertDialog` and on `Button`.
```
Error:(25, 34) TS2322: Type 'MutableRefObject<HTMLButtonElement | FocusableElement | undefined>' is not assignable to type 'RefObject<FocusableElement>'.
  Types of property 'current' are incompatible.
    Type 'HTMLButtonElement | FocusableElement | undefined' is not assignable to type 'FocusableElement | null'.
      Type 'undefined' is not assignable to type 'FocusableElement | null'.


Error:(37, 21) TS2322: Type 'MutableRefObject<HTMLButtonElement | FocusableElement | undefined>' is not assignable to type 'string | ((instance: HTMLButtonElement | null) => void) | RefObject<HTMLButtonElement> | null | undefined'.
  Type 'MutableRefObject<HTMLButtonElement | FocusableElement | undefined>' is not assignable to type 'RefObject<HTMLButtonElement>'.
    Types of property 'current' are incompatible.
      Type 'HTMLButtonElement | FocusableElement | undefined' is not assignable to type 'HTMLButtonElement | null'.
        Type 'undefined' is not assignable to type 'HTMLButtonElement | null'.
```

BTW, also having a typescript section in documentation will be nice. For some hard to figure out use-cases. Like creating your own components from chakra primitives etc.

**Comment by ljosberinn** [comment, marked as answer]
Could you provide a codesandbox and your TS version? Works fine here: https://codesandbox.io/s/musing-glade-n3zns?file=/src/App.tsx

**Reply by RIP21** [reply]
I guess this is how it's supposed to be. You added `null` inside `useRef` which is a counter-intuitive fix, but it works :) 
https://codesandbox.io/s/stoic-edison-wh8m9?file=/src/index.tsx

Passing `null` into `useRef` explicitly removes warnings. I don't remember anyone recommending setting `useRef` to null in any of React docs (maybe they are) :) While it changes almost nothing really, it fixes error...  And what's most important, regular native HTML `button` also complains without explicit `null`. So I guess today I learned?

Thanks, me being a hardcore user of TS for almost 2 years doesn't stop me from doing stupid mistakes like that one :D

**Reply by ljosberinn** [reply]
The error there is:
```ts
Type 'MutableRefObject<HTMLButtonElement | undefined>' is not assignable to type 'string | ((instance: HTMLButtonElement | null) => void) | RefObject<HTMLButtonElement> | null | undefined'.
  Type 'MutableRefObject<HTMLButtonElement | undefined>' is not assignable to type 'RefObject<HTMLButtonElement>'.
    Types of property 'current' are incompatible.
      Type 'HTMLButtonElement | undefined' is not assignable to type 'HTMLButtonElement | null'.
        Type 'undefined' is not assignable to type 'HTMLButtonElement | null'.
```
so apparently useRef is somehow typed as `T | null` internally, hence `null` removes the warning. Tbh regardless I'm personally always prefering `null` over `undefined` because `undefined` is implicitly used by JS already, while `null` must come from my (or any library) code.

**Reply by RIP21** [reply]
I thought that `current` is `null` by default if you're not passing anything to it and because of that, it typed as `T | null` while `current` can be reassigned to `undefined` manually after, I guesed, by pure `ref.current = undefined`, but the truth is, `ref.current` will be `undefined` if you pass nothing to `useRef` hook. 
So when you do not pass anything in it, the compiler just correctly thinks that type would be `T | null | undefined` letting you set three of those to `current`, while nobody does that, I think it's probably correct typing anyway and just how TS infers generics from arguments.

Also, in the official doc they also pass `null` explicitly :) 
https://reactjs.org/docs/hooks-reference.html#useref

BTW, who did typings for chakra since version 1.0? It's purely awesome. I mean, `as` prop magic without generics it just WOW and I want to understand how it works, as it's an inference on another level.
The only issue tho is that Webstorm/IntelliJ has no autocomplete whatsoever in JSX for some reason for Chakra, but the compiler slaps in the face in case of issues, so it's all good, just smth with the IDE itself :)

**Reply by ljosberinn** [reply]
- not passing an argument is effectively `useRef(undefined)` :)
- typings for v1 were mostly done by Segun with additional work from Tim especially recently - [here's a PR that was meant to fix the issues for IntelliJ](https://github.com/chakra-ui/chakra-ui/pull/2753) but apparently there's also an internal problem with Pick and/or Omit within IntelliJ so technically an upstream issue

**Reply by RIP21** [reply]
@ljosberinn sure :) That's obvious :) 

Regarding IntelliJ. 
Yeah the issue is opened at their bug tracker, so let's hope it will be fixed from their side :)

**Reply by ovirta** [reply]
Is this something that should be covered in the example as well, where null is not passed to useRef?

https://chakra-ui.com/docs/components/overlay/alert-dialog

**Reply by aaarichter** [reply]
@ovirta most definitely - it would be great if all chakra code examples had a typescript version.
